### PR TITLE
Adding promises and operation timeouts to fix race conditions and infinite loading in OE

### PR DESCRIPTION
**Pull request by aasimkhan30**
There are 4 bugs that I am trying to solve here 
1. The flow of ads create session is as follow 

ads ----- createNewSession ---> provider
ads <----createNewSessionResponse----- provider
ads <----onSessionCreated-------- provider

Occasionally, there are situations where ads receives the 2 responses from provider in reverse order which messes up the session state within ads. I am making sure that we get createNewSessionResponse for the connection before proceeding to run code for onSessionCreated.

2. Occasionally providers like PGSQL and Mysql will return malformed 'createNewSesssionResponse' thereby messing up the connection state in ads. The PR fixes that bug too by making sure the response is not malformed and necessary props are present in the response.

3.  In certain cases, when we try to connect to a previously connected server, we may encounter a situation where the session is present in this._sessions, but the connection is outdated. This happens when the handleSessionCreated is recieved before the createNewSession response is received.  We are using a map to keep track of connections which are waiting for a session to be created. And we will continue only if we get createNewSession response for those connections.

4. Occasionally in async tree we keep waiting for provider to return nodes but never get them. This causes infinite loading in async tree. This PR fixes that too.

### CODE REVIEW

File: src/sql/workbench/services/objectExplorer/browser/treeUpdateUtils.ts [hunk 1]
```diff
@@ -274,21 +287,40 @@ export class TreeUpdateUtils {
 				await objectExplorerService.resolveTreeNodeChildren(session, rootNode);
 				return rootNode.children ?? [];
 			} else {
+				let nodesUpdatedListener;
```
Comment (kburtram) [inline src/sql/workbench/services/objectExplorer/browser/treeUpdateUtils.ts:290]: why no declare inside the promise, for example where it's defined?

**Review by kburtram — APPROVED** [review]

File: src/sql/workbench/services/objectExplorer/browser/asyncServerTreeDataSource.ts [hunk 2]
```diff
@@ -45,25 +47,19 @@ export class AsyncServerTreeDataSource implements IAsyncDataSource<ConnectionPro
...
-			if (element instanceof TreeNode) {
-				element.errorStateMessage = err.message ?? err;
-			}
-			if (err.message) {
-				this.showError(err.message);
-			}
-
-			throw err;
+		} catch (error) {
+			this.showError(error);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/asyncServerTreeDataSource.ts:61]: This doesn't seem related to your description of the changes - why did you remove the TreeNode check?

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts [hunk 3]
```diff
@@ -276,8 +279,64 @@ export class ObjectExplorerService implements IObjectExplorerService {
...
+				} else if (count > this.getObjectExplorerTimeout()) {
+					clearTimeout(timerId);
+					reject(new Error(
+						nls.localize(
+							'objectExplorerMissingSession',
+							'Timed out while waiting for createNewSession response from the provider for this created session `{0}`', sessionId
+						)));
+				} else {
+					count++;
+					timerId = setTimeout(waitingForSessionId, 1000);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:304]: We should really avoid poll based timeouts like this. It would be better to have a single timeout for the overall OE timeout occurring, and then cancelling that if we get a response before that.

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts [hunk 4]
```diff
@@ -297,13 +356,13 @@ export class ObjectExplorerService implements IObjectExplorerService {
...
 			} catch (error) {
-				this.logService.warn(`cannot handle the session ${session.sessionId} in all nodeProviders`);
+				this.logService.warn(`cannot handle the session ${sessionId} in all nodeProviders`);
 			} finally {
 				this.sendUpdateNodeEvent(connection!, errorMessage);
 			}
 		}
 		else {
-			this.logService.warn(`cannot find session ${session.sessionId}`);
+			this.logService.warn(`cannot find session ${sessionId}`);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:365]:
```suggestion
			this.logService.warn(`Cannot find session ${sessionId}`);
```

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts [hunk 5]
```diff
@@ -276,8 +279,64 @@ export class ObjectExplorerService implements IObjectExplorerService {
...
+				const session = this._sessions[sessionId];
+				if (session) {
+					clearTimeout(timerId);
+					resolve();
+				} else if (count > this.getObjectExplorerTimeout()) {
+					clearTimeout(timerId);
+					reject(new Error(
+						nls.localize(
+							'objectExplorerMissingSession',
+							'Timed out while waiting for createNewSession response from the provider for this created session `{0}`', sessionId
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:300]: We don't really use backticks for quotes anywhere else so that should be avoided - and here the session ID isn't going to contain any spaces so is fine to just have not quoted I think. 

```suggestion
							'Timed out while waiting for createNewSession response from the provider for this created session {0}', sessionId
```

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts [hunk 6]
```diff
@@ -276,8 +279,64 @@ export class ObjectExplorerService implements IObjectExplorerService {
...
+					} else if (count > this.getObjectExplorerTimeout()) {
+						clearTimeout(timerId);
+						reject(new Error(
+							nls.localize(
+								'objectExplorerMissingConnectionForSession',
+								'Timeout waiting for session "{0}" to be created for connection "{0}"', connection.id, sessionId
+							)));
+					} else {
+						count++;
+						timerId = setTimeout(waitingForRightConnectionInSession, 1000);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:335]: Same here about avoiding poll-based timeouts.

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts [hunk 7]
```diff
@@ -360,16 +419,26 @@ export class ObjectExplorerService implements IObjectExplorerService {
...
+			}
 			if (this._sessions[result.sessionId]) {
 				this.logService.trace(`Overwriting session ${result.sessionId}`);
 			}
 			this._sessions[result.sessionId] = {
 				connection: connection,
 				nodes: {}
 			};
+			// once the session is created, set the connection to not wait for session
+			this._connectionsWaitingForSession.set(connection.id, false);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:438]: If you store a deferred promise here instead of just a boolean then you can use that to complete the promise above in place of the polling timeout (or use an event)

File: src/sql/workbench/services/objectExplorer/browser/treeUpdateUtils.ts [hunk 8]
```diff
@@ -281,14 +294,31 @@ export class TreeUpdateUtils {
...
+					}
+					const nodeUpdateTimer = setTimeout(nodeUpdateTimeout, expansionTimeoutValueSec * 1000);
+
+
+					const nodesUpdatedListener = objectExplorerService.onUpdateObjectExplorerNodes(e => {
+						if (e.connection && e.connection.id === connection.id) {
+							if (e.errorMessage) {
+								reject(new Error(e.errorMessage));
+							} else {
+								resolve(undefined);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/treeUpdateUtils.ts:319]: If you specify this as a `Promise<void>` then you can get rid of the undefined here - since you're not using the return value anyways.

**Review by lewis-sanchez — APPROVED** [review]

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts [hunk 9]
```diff
@@ -276,35 +281,80 @@ export class ObjectExplorerService implements IObjectExplorerService {
...
+					this._connectionsWaitingForSession.delete(connection.id);
 				}
-				else {
-					errorMessage = session && session.errorMessage ? session.errorMessage : errSessionCreateFailed;
-					this.logService.error(errorMessage);
+				createNewSessionListener.dispose();
+				clearTimeout(timeout);
+			}
+			const onTimeout = () => {
+				if (!this._sessions[sessionId]) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:296]: Do we actually need separate errors here? Seems like the stuff in the else is all we need

File: src/sql/workbench/services/objectExplorer/browser/treeUpdateUtils.ts [hunk 10]
```diff
@@ -281,14 +294,31 @@ export class TreeUpdateUtils {
...
+				const nodesUpdatedPromise = new Promise<void>((resolve, reject) => {
+					// Clean up timeout and listener
+					const cleanup = () => {
+						clearTimeout(nodeUpdateTimer);
+						nodesUpdatedListener.dispose();
+					}
+
+					// If the node update takes too long, reject the promise
+					const nodeUpdateTimeout = () => {
+						reject(new Error(nls.localize('objectExplorerTimeout', "Object Explorer timed out for {0}", connection.databaseName)));
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/treeUpdateUtils.ts:308]:
```suggestion
						reject(new Error(nls.localize('objectExplorerTimeout', "Object Explorer expansion timed out for '{0}'", connection.databaseName)));
```

**Review by Charles-Gagnon — APPROVED** [review]
It would be nice to have some new tests added as well

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts (hunk 9 above)
Comment (aasimkhan30) [reply src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:296]: Usually, this error won't happen unless the providers are misbehaving. Having separate errors will just help us figure out what exactly is happening in the OE service.

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts [hunk 11]
```diff
@@ -276,35 +281,80 @@ export class ObjectExplorerService implements IObjectExplorerService {
...
+					this.logService.error(`Timed out waiting for session to be created`);
+					reject(new Error('Timed out waiting for session to be created'));
+				} else {
+					this.logService.error(`Timeout waiting for session ${sessionId} to be created for connection "${connection.title}"`);
+					reject(new Error(nls.localize(
+						'objectExplorerMissingConnectionForSession',
+						'Timeout waiting for session {0} to be created for connection "{0}"', sessionId, connection.title
```
Comment (alanrenmsft) [inline src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:303]: here it is localized, definitely should refactor the code to reuse this branch.

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts [hunk 12]
```diff
@@ -276,35 +281,80 @@ export class ObjectExplorerService implements IObjectExplorerService {
...
 				}
-				// Send on session created about the session to all node providers so they can prepare for node expansion
-				let nodeProviders = this._nodeProviders[connection!.providerName];
-				if (nodeProviders) {
-					const promises = nodeProviders.map(p => p.handleSessionOpen(session));
-					await Promise.all(promises);
+
+				cleanup();
+			}
+			const timeout = setTimeout(onTimeout, this.getObjectExplorerTimeout() * 1000);
```
Comment (alanrenmsft) [inline src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:309]: rename to timeoutHandle to make it clear.

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts [hunk 13]
```diff
@@ -276,35 +281,80 @@ export class ObjectExplorerService implements IObjectExplorerService {
...
+				this.logService.error(errorMessage);
+			}
+			// Send on session created about the session to all node providers so they can prepare for node expansion
+			let nodeProviders = this._nodeProviders[connection!.providerName];
+			if (nodeProviders) {
+				const promises = nodeProviders.map(p => p.handleSessionOpen(session));
+				await Promise.all(promises);
+			}
+		} catch (error) {
+			this.logService.warn(`cannot handle the session ${sessionId} in all nodeProviders`);
```
Comment (alanrenmsft) [inline src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:353]: the actual error message is ignored here, don't you need it to troubleshoot?

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts (hunk 9 above)
Comment (Charles-Gagnon) [reply src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:296]: Please update the messages to be clearer though - right now they don't really indicate the problem at all. For example, you always have a sessionID so should always be including that. And then neither branch does anything with the actual session or lack thereof - the second adds some info about the connection but isn't that also available in both cases as well?

Comment (aasimkhan30) [reply src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:296]: @Charles-Gagnon , updated the error message to include more context.

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts [hunk 14]
```diff
@@ -276,35 +281,84 @@ export class ObjectExplorerService implements IObjectExplorerService {
...
+				createNewSessionListener.dispose();
+				clearTimeout(timeoutHandle);
+			}
+			const onTimeout = () => {
+				if (!this._sessions[sessionId]) {
+					this.logService.error(`Timed out waiting for session ${sessionId} to be created.
+					This has probably happened because OE service did not recieve a response for createNewSession from the provider.`);
+					reject(new Error(
+						nls.localize('objectExplorerMissingSession',
+							'Timed out waiting for session {0} to be created. This has probably happened because OE service did not recieve a response for createNewSession from the provider.', sessionId)));
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:301]: The user isn't going to know what the session ID means, and the reasoning why isn't something they would understand either so I think we should just simplify it.

```suggestion
							'Timed out waiting for the Object Explorer session to be created.')));
```

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts [hunk 15]
```diff
@@ -276,35 +281,84 @@ export class ObjectExplorerService implements IObjectExplorerService {
...
+				} else {
+					this.logService.error(`Timeout waiting for session ${sessionId} to be created for connection "${connection.title}".
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:303]: I'm still confused about this else. You say it timed out waiting for the session to be created - but you were able to get a value back from `this._sessions[sessionId]`? Doesn't that imply that a session was found so it succeeded?

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts (hunk 13 above)
Comment (aasimkhan30) [reply src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts]: Good catch. Updated the error message to include the error as well.

File: src/sql/workbench/contrib/objectExplorer/common/serverGroup.contribution.ts [hunk 16]
```diff
@@ -41,6 +41,7 @@ const serverGroupConfig: IConfigurationNode = {
 	}
 };
 
+export const NODE_EXPANSION_CONFIG = 'serverTree.nodeExpansionTimeout';
 const serverTreeConfig: IConfigurationNode = {
 	'id': 'serverTree',
 	'title': 'Server Tree',
```
Comment (alanrenmsft) [inline src/sql/workbench/contrib/objectExplorer/common/serverGroup.contribution.ts:47]: while you are here, please also localize these strings.

**Review by alanrenmsft — APPROVED** [review]

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts (hunk 15 above)
Comment (aasimkhan30) [reply src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:303]: This bug happens when I am frequently connecting and disconnecting a connection.

This code gets executed whenever we disconnect a OE connection.
https://github.com/microsoft/azuredatastudio/blob/30c22a475ef1a0c51d9eefd919cf06693acb15f0/src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts#L236-L238
Sometimes, close session takes some time to finish. When that happens, this_sessions[sessionId] still has connection property from the previous session.

While we are waiting for closeSession, if I create a new session, and handleSession is called before we get the response of createNewSession this piece of code is not executed..
https://github.com/microsoft/azuredatastudio/blob/30c22a475ef1a0c51d9eefd919cf06693acb15f0/src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts#L428-L431

Many providers reuse session ids and therefore the value of 
this._session[sessionId].connection is from the old connection and not the new one.

Comment (aasimkhan30) [reply src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:303]: @Charles-Gagnon , we can sync up offline if the explanation is not very clear.

Comment (Charles-Gagnon) [reply src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:303]: So it sounds like we don't clear the map until after getting the close session response? 

If that's the case then we should maybe look at clearing the map BEFORE requesting the session be closed - basically the idea being that the user indicated they want to close the session so we should do so on the client side as much as we can and then do a best effort to get the provider to clean the session up themselves as well (which may fail due to other issues, but that shouldn't block us from cleaning up our stuff internally)

Anyways, if that does is something you want to follow up on then I'd suggest creating an issue for it and sending out a separate PR since that's existing behavior and outside the scope of this PR. 

You may want to still consider at least adding a comment explaining this though - it's doesn't seem obvious to me and documenting that may help someone else investigating this in the future.

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts [hunk 17]
```diff
@@ -276,35 +281,84 @@ export class ObjectExplorerService implements IObjectExplorerService {
...
+					This has probably happened because OE service did not recieve a response for createNewSession from the provider.`);
+					reject(new Error(
+						nls.localize('objectExplorerMissingSession',
+							'Timed out waiting for the Object Explorer session to be created.')));
+				} else {
+					this.logService.error(`Timeout waiting for session ${sessionId} to be created for connection "${connection.title}".
+					This has probably happened because OE service did not recieve a response for createNewSession from the provider for connection."${connection.title}`);
+					reject(new Error(nls.localize(
+						'objectExplorerMissingConnectionForSession',
+						'Timeout waiting for session {0} to be created for connection "{1}". This has probably happened because OE service did not recieve a response for createNewSession from the provider for connection "{1}"', sessionId, connection.title
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:307]:
```suggestion
						"Timed out waiting for the Object Explorer session to be created for connection '{1}'.", connection.title
```

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts [hunk 18]
```diff
@@ -360,16 +414,27 @@ export class ObjectExplorerService implements IObjectExplorerService {
 	public async createNewSession(providerId: string, connection: ConnectionProfile): Promise<azdata.ObjectExplorerSessionResponse> {
 		const provider = this._providers[providerId];
 		if (provider) {
+			// set the connection to wait for session
+			this._connectionsWaitingForSession.set(connection.id, true);
 			const result = await provider.createNewSession(connection.toConnectionInfo());
+			// some providers return a malformed create sessions responses which don't have a session id. We should throw an error in this case
+			if (!result?.sessionId) {
+				this.logService.error(`The session ID returned by provider "${providerId}" for connection "${connection.title}" is invalid.`);
+				throw new Error(nls.localize('objectExplorerSessionIdMissing', 'The session ID returned by provider "{0}" for connection "{1}" is invalid.', providerId, connection.title));
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:423]: nit: But generally we've been using single quotes for quoting text displayed to users

```suggestion
				throw new Error(nls.localize('objectExplorerSessionIdMissing', "No session ID returned by provider '{0}' for connection '{1}'.", providerId, connection.title));
```

File: src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts [hunk 19]
```diff
@@ -360,16 +414,27 @@ export class ObjectExplorerService implements IObjectExplorerService {
 	public async createNewSession(providerId: string, connection: ConnectionProfile): Promise<azdata.ObjectExplorerSessionResponse> {
 		const provider = this._providers[providerId];
 		if (provider) {
+			// set the connection to wait for session
+			this._connectionsWaitingForSession.set(connection.id, true);
 			const result = await provider.createNewSession(connection.toConnectionInfo());
+			// some providers return a malformed create sessions responses which don't have a session id. We should throw an error in this case
+			if (!result?.sessionId) {
+				this.logService.error(`The session ID returned by provider "${providerId}" for connection "${connection.title}" is invalid.`);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts:422]:
```suggestion
				this.logService.error(`No session ID returned by provider '${providerId}' for connection '${connection.title}'.`);
```

**Review by Charles-Gagnon — APPROVED** [review]

**Review by cheenamalhotra — APPROVED** [review]
